pytensor/scalar/math.py

```diff
@@ -1481,3 +1481,155 @@ def c_code(self, *args, **kwargs):
 
 
 betainc_der = BetaIncDer(upgrade_to_float_no_complex, name="betainc_der")
+
+
+class Hyp2F1(ScalarOp):
+    """
+    Gaussian hypergeometric function ``2F1(a, b; c; z)``.
+
+    """
+
+    nin = 4
+    nfunc_spec = ("scipy.special.hyp2f1", 4, 1)
+
+    @staticmethod
+    def st_impl(a, b, c, z):
+
+        if abs(z) >= 1:
+            raise NotImplementedError("hyp2f1 only supported for z < 1.")
+        else:
+            return scipy.special.hyp2f1(a, b, c, z)
+
+    def impl(self, a, b, c, z):
+        return Hyp2F1.st_impl(a, b, c, z)
+
+    def grad(self, inputs, grads):
+        a, b, c, z = inputs
+        (gz,) = grads
+        return [
+            gz * hyp2f1_der(a, b, c, z, 0),
+            gz * hyp2f1_der(a, b, c, z, 1),
+            gz * hyp2f1_der(a, b, c, z, 2),
+            gz * hyp2f1_der(a, b, c, z, 3),
+        ]
+
+    def c_code(self, *args, **kwargs):
+        raise NotImplementedError()
+
+
+hyp2f1 = Hyp2F1(upgrade_to_float, name="hyp2f1")
+
+
+class Hyp2F1Der(ScalarOp):
+    """
+    Derivatives of the Gaussian hypergeometric function ``2F1(a, b; c; z)``.
+
+    """
+
+    nin = 5
+
+    def impl(self, a, b, c, z, wrt):
+        def _infinisum(f):
+            """
+            Utility function for infinite summations.
+            """
+
+            n, res = 0, f(0)
+            while True:
+                term = f(n + 1)
+                if RuntimeWarning:
+                    break
+                if (res + term) - res == 0:
+                    break
+                n, res = n + 1, res + term
+            return res
+
+        def _hyp2f1_da(a, b, c, z):
+            """
+            Derivative of hyp2f1 wrt a
+
+            """
+
+            if abs(z) >= 1:
+                raise NotImplementedError("Gradient not supported for |z| >= 1")
+
+            else:
+                term1 = _infinisum(
+                    lambda k: (
+                        (gamma(a + k) / gamma(a))
```

Comment:
On another note, these functions seem to really want to be computed on the log scale for numerical stability

Reply:
After rewriting with scipy I see the tests failing with `nan`, suggesting the numerical stability I mentioned matters or your test inputs are too wild

Reply:
`scipy` functions are now being used. Could log scale computations possibly fix this new error?

```
ValueError: ('abs_err not finite', 'array([[nan, nan, nan],\n       [nan, nan, nan]])', 'Test Elemwise{hyp2f1,no_inplace}::normal: 
Error occurred while computing the gradient on the following inputs: [array([[764.16214925, 550.49823533, 
542.19109217],\n       [613.93532095, 341.6967123 , 284.38215306]]), array([[764.16214925, 550.49823533, 
542.19109217],\n       [613.93532095, 341.6967123 , 284.38215306]]), array([[764.16214925, 550.49823533, 
542.19109217],\n       [613.93532095, 341.6967123 , 284.38215306]]), array([[0.38208107, 0.27524912, 0.27109555],\n
[0.30696766, 0.17084836, 0.14219108]])]')
```

Is this as simple as adding an `ln` to the function terms and an `exp` to the returned value?

Reply:
If rewriting these derivatives in Stan is a Herculean effort, do you think it would be more prudent to proceed with using a `scan` implementation with the added helper functions for `hyp2f1`? The derivatives would not need to be provided in this case, and this would be done in a future PR anyway. It may not be the most performant option rightt now, but I don't see `hyp2f1` being used for anything other than the upcoming ParetoNBD model.

Here's a rough stab at it:

```
def converge(prev_result, result):
    return prev_result, aesara.scan.utils.until(prev_result==result)
            
results, updates = aesara.scan(
    fn = lambda n, a, b, c, z: if abs(z) >=1 raise NotImplementedError() else ((poch(a,n) * poch(b,n))/poch(c,n)) * ((z**n)/factorial(n)),
        outputs_info=at.constant(1.),
        non_sequences=[a,b,c,z],
        n_steps = 1024
        )
            
        final_result = results.sum()

hyp2f1 = aesara.function(inputs=[a,b,c,z], outputs=final_result , updates=updates)
```